doc/source/rllib-concepts.rst

```diff
@@ -346,6 +346,37 @@ In PPO we run ``setup_mixins`` before the loss function is called (i.e., ``befor
 
 Finally, note that you do not have to use ``build_tf_policy`` to define a TensorFlow policy. You can alternatively subclass ``Policy``, ``TFPolicy``, or ``DynamicTFPolicy`` as convenient.
 
+Building Policies in TensorFlow Eager
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+While RLlib runs all TF operations in graph mode, you can still leverage TensorFlow eager using `tf.py_function <https://www.tensorflow.org/api_docs/python/tf/py_function>`__. However, note that eager and non-eager tensors cannot be mixed within the ``py_function``. Here's an example of embedding eager execution within a policy loss function:
+
+.. code-block:: python
+
+    def eager_loss(policy, batch_tensors):
+        """Example of using embedded eager execution in a custom loss.
+
+        Here `compute_penalty` prints the actions and rewards for debugging, and
+        also computes a (dummy) penalty term to add to the loss.
+        """
+
+        def compute_penalty(actions, rewards):
+            penalty = tf.reduce_mean(tf.cast(actions, tf.float32))
+            if random.random() > 0.9:
+                print("The eagerly computed penalty is", penalty, actions, rewards)
+            return penalty
+
+        actions = batch_tensors[SampleBatch.ACTIONS]
+        rewards = batch_tensors[SampleBatch.REWARDS]
+        penalty = tf.py_function(
+            compute_penalty, [actions, rewards], Tout=tf.float32)
```

Comment:
So is the trick mainly to run py_functions in the static mode?

Reply:
Yep. Right now this is limited to loss_fn, but with ModelV2 this can also include the model forward pass.